omisego-sdk/src/main/java/co/omisego/omisego/custom/retrofit2/converter/OMGConverterFactory.kt

```diff
@@ -0,0 +1,69 @@
+package co.omisego.omisego.custom.retrofit2.converter
+
+/*
+ * OmiseGO
+ *
+ * Created by Phuchit Sirimongkolsathien on 5/3/2018 AD.
+ * Copyright © 2017-2018 OmiseGO. All rights reserved.
+ */
+
+import co.omisego.omisego.constant.ErrorCode
+import co.omisego.omisego.constant.Versions
+import co.omisego.omisego.exception.OMGAPIErrorException
+import co.omisego.omisego.model.APIError
+import co.omisego.omisego.model.OMGResponse
+import com.google.gson.Gson
+import com.google.gson.JsonElement
+import com.google.gson.TypeAdapter
+import com.google.gson.reflect.TypeToken
+import okhttp3.ResponseBody
+import org.json.JSONException
+import retrofit2.Converter
+import retrofit2.Retrofit
+import java.io.ByteArrayInputStream
+import java.io.IOException
+import java.io.InputStreamReader
+import java.lang.reflect.Type
+
+
+internal class OMGConverterFactory(private val gson: Gson) : Converter.Factory() {
+
+    companion object {
+        fun create(gson: Gson): OMGConverterFactory {
+            return OMGConverterFactory(gson)
+        }
+    }
+
+    override fun responseBodyConverter(type: Type, annotations: Array<Annotation>, retrofit: Retrofit): Converter<ResponseBody, *> {
+        val adapter = gson.getAdapter(TypeToken.get(type))
+        return OMGConverter(gson, adapter)
+    }
+
+    internal class OMGConverter<T>(private val gson: Gson, private val adapter: TypeAdapter<T>) : Converter<ResponseBody, T> {
+        @Throws(IOException::class)
+        override fun convert(responseBody: ResponseBody): T {
```

Comment:
I think we can optimize this function, I'll check later and open a PR if it's actually the case.
It's not a blocker anyway :)

Reply:
I'm already done this :)

Reply:
Nice improvements, but you can do even better:
- If you check the implementation of `responseBody.string()` it doesn't cache the result, so better do it yourself in a var
- There's no need to create the `inputStream` and `reader` I believe, you can do directly:

```kotlin
                /* Initialize JSONObject from plain response string */
                val responseRaw = responseBody.string()
                val response = JSONObject(responseRaw)

                /* Get the success flag */
                val success = response.getBoolean("success")

                return when {
                    success -> adapter.fromJson(responseRaw)
                    else -> {
                        val errorToken = object : TypeToken<OMGResponse<APIError>>() {}.type
                        val error = gson.fromJson<OMGResponse<APIError>>(responseRaw, errorToken)
                        throw OMGAPIErrorException(error)
                    }
                }
```

Btw, there's still a potential perf issue here, we're parsing the json **twice** (the first time as a JSONObject, the second time as the real type. In the future it would be interesting to benchmark this vs doing something like

- Parse `responseBody.string()` as a `OMGResponse<String>` (gson is capable of keeping json object as string, so `data` would be the json as string of the actual `data` field)
- Check if success or not, and decide the real type argument of `OMGResponse` (`APIError` or `User, Balance` ...)
- Parse the `data` of the above `OMGResponse<String>` as the actual type
- Create a new `OMGResponse` based on the existing one by just replacing the `data` property with the parsed value

The issue with this is that your `T` type parameter in the converter is in fact `OMGResponse<User>` and not `User`, so you'd have to extract it.
Plus this makes you add a constructor in OMGResponse like:

```
constructor(copy: OMGResponse<*>, data: T) : this(
            version = copy.version,
            success = copy.success,
            data = data
    )
```

So you would have to maintain this constructor as you update the primary one, but since it's in the same file it's ok.
But as I said, not sure it's worth the trouble, it would need some benchmarking one day :)

In the  mean time, the first suggestion should still be apply to simplify :)

Reply:
I removed the `InputStreamReader`. Btw, your concern about perf is interesting that which way is better. However, I think that should not be our top priority since we have to do some research first by writing another way and doing a perf test (you can do researching and send me a PR if you think it should be done now). We should focus on making thing working fine. There're still have more features to do in the SDK like `Listing the transaction`, `Create the QR code`, etc 😄

Reply:
Yes! Totally agree! 👌